src/fsharp/block.fs

```diff
@@ -0,0 +1,222 @@
+
+            let builder = ImmutableArray.CreateBuilder(n)
+            for i = 0 to n - 1 do
+                builder.Add(f i)
+            builder.ToImmutable()
+
+    let iter f (arr: block<'T>) =
+        for i = 0 to arr.Length - 1 do
+            f arr.[i]
+
+    let iteri f (arr: block<'T>) =
+        for i = 0 to arr.Length - 1 do
+            f i arr.[i]
+
+    let iter2 f (arr1: block<'T1>) (arr2: block<'T2>) =
+        if arr1.Length <> arr2.Length then
+            invalidOp "Block lengths do not match."
+
+        for i = 0 to arr1.Length - 1 do
+            f arr1.[i] arr2.[i]
+
+    let iteri2 f (arr1: block<'T1>) (arr2: block<'T2>) =
+        if arr1.Length <> arr2.Length then
+            invalidOp "Block lengths do not match."
+
+        for i = 0 to arr1.Length - 1 do
+            f i arr1.[i] arr2.[i]
+
+    let map (mapper: 'T -> 'U) (arr: block<'T>) : block<_> =
+        match arr.Length with
+        | 0 -> block.Empty
+        | 1 -> ImmutableArray.Create(mapper arr.[0])
+        | _ ->
+            let builder = ImmutableArray.CreateBuilder(arr.Length)
+            for i = 0 to arr.Length - 1 do
+                builder.Add(mapper arr.[i])
+            builder.ToImmutable()
+
+    let mapi (mapper: int -> 'T -> 'U) (arr: block<'T>) : block<_> =
+        match arr.Length with
+        | 0 -> block.Empty
+        | 1 -> ImmutableArray.Create(mapper 0 arr.[0])
+        | _ ->
+            let builder = ImmutableArray.CreateBuilder(arr.Length)
+            for i = 0 to arr.Length - 1 do
+                builder.Add(mapper i arr.[i])
+            builder.ToImmutable()
+
+    let map2 (mapper: 'T1 -> 'T2 -> 'T) (arr1: block<'T1>) (arr2: block<'T2>) : block<_> =
+        if arr1.Length <> arr2.Length then
+            invalidOp "Block lengths do not match."
+      
+        match arr1.Length with
+        | 0 -> ImmutableArray.Empty
+        | 1 -> ImmutableArray.Create(mapper arr1.[0] arr2.[0])
+        | n ->
+            let builder = ImmutableArray.CreateBuilder(n)
+            for i = 0 to n - 1 do
+                builder.Add(mapper arr1.[i] arr2.[i])
+            builder.ToImmutable()
+
+    let mapi2 (mapper: int -> 'T1 -> 'T2 -> 'T) (arr1: block<'T1>) (arr2: block<'T2>) : block<_> =
+        if arr1.Length <> arr2.Length then
+            invalidOp "Block lengths do not match."
+      
+        match arr1.Length with
+        | 0 -> ImmutableArray.Empty
+        | 1 -> ImmutableArray.Create(mapper 0 arr1.[0] arr2.[0])
+        | n ->
+            let builder = ImmutableArray.CreateBuilder(n)
+            for i = 0 to n - 1 do
+                builder.Add(mapper i arr1.[i] arr2.[i])
+            builder.ToImmutable()
+
+    let concat (arr: block<block<'T>>) : block<'T> =
+        if arr.IsEmpty then
+            empty
+        elif arr.Length = 1 then
+            arr.[0]
+        elif arr.Length = 2 then
+            arr.[0].AddRange(arr.[1])
+        else
+            let mutable arr2 = arr.[0]
```

Comment:
Maybe best to do this with a builder with a decent starting capacity, lots of copies here otherwise.

Reply:
Yes, you are right about that.